dgisim/src/dices.py

```diff
@@ -162,6 +175,205 @@ def just_satisfy(self, requirement: AbstractDices) -> bool:
+                                   else local_precedence[0])  # list of first priority elements
+
+        def get_precedence(element: Element) -> Tuple[int, int]:
+            nonlocal local_precedence
+            """get precedence by element, bigger number is lower precedence"""
+            for i, set_ in enumerate(local_precedence, start=1):
+                if element in set_:
+                    return i, ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY[::-1].index(
+                        element)
+            else:
+                return BIG_INT, 0
+
+        def check_asserts():
+            """
+            assert non-negativity of variables:
+            need supply result_dct
+            """
+
+            def check_dict(name: str, dct: Dict):
+                for el, val in dct.items():
+                    assert val >= 0, f"{el} in {name} < 0 {val=}"
+
+            nonlocal result_dct, supply, need
+
+            check_dict("result_dct", result_dct)
+            check_dict("supply", supply)
+            check_dict("need", need)
+
+        # 1 step - fill OMNI requirement
+        if requirement[Element.OMNI] > 0:
+            def comparision(x: Tuple[Element, int]) -> Tuple[int, int, int]:
+                """to sort by ascedence of least_spend"""
+                type_of_element: Element = x[0]
+                number_of_dices: Optional[int] = x[1]
+                priority_local, priority_global = (
+                    get_precedence(type_of_element))
+                # return number_of_dices if number_of_dices is not None else
+                # -1, type_of_element
+                return number_of_dices, priority_local, priority_global
+
+            array: List[Tuple[Element, Optional[int]]] = \
+                [(element,
+                  self._how_much_omni_and_first_priority_spends(
+                      element_supply=supply[element],
+                      element_need=need[element],
+                      omni_need=need[Element.OMNI],
+                      element_priority=element in first_priority_elements,
+                      omni_supply=supply[Element.OMNI])
+                  )
+                 for element in _PURE_ELEMS]
+
+            # TODO rewrite minimum get
+            array = list(sorted(array, key=comparision))
+
+            least_spend = array[0][1]
+            least_spend_omni_elements = []
+            if least_spend != BIG_INT:
+                for el in array:
+                    if el[1] == least_spend:
+                        least_spend_omni_elements.append(el[0])
+            else:
+                raise Exception("Cant feed OMNI requirement")
+
+            assert least_spend_omni_elements, "Unknown error"
+
+            precedence_of_least_spend_omni = {element: get_precedence(element)
+                                              for element
+                                              in least_spend_omni_elements}
```

Comment:
this should be list[tuple[Element, int]] instead of a dictionary if you are not using the features of a dictionary. (when I say feature, I mean "random access" by keys.